src/ppom-storage.ts

```diff
@@ -0,0 +1,208 @@
+ */
+export type PPOMFileMetadata = FileInfo[];
+
+/**
+ * StorageKey Type
+ * This defines type of key that is used for indexing file data saved in StorageBackend.
+ */
+export type StorageKey = {
+  name: string;
+  chainId: string;
+};
+
+/**
+ * StorageBackend Type
+ * This defines type for storage backend implementation.
+ * There will be different storage implementations depending on platform:
+ * 1. extension - indexDB
+ * 2. mobile app - <TBD>
+ */
+export type StorageBackend = {
+  read(key: StorageKey, checksum: string): Promise<ArrayBuffer>;
+  write(key: StorageKey, data: ArrayBuffer, checksum: string): Promise<void>;
+  delete(key: StorageKey): Promise<void>;
+  dir(): Promise<StorageKey[]>;
+};
+
+/**
+ * PPOMStorage class
+ * This class is responsible for managing the local storage
+ * It provides the following functionalities:
+ * 1. Sync the metadata with the version info from the cdn
+ * 2. Read a file from the local storage
+ * 3. Write a file to the local storage
+ *
+ * It also validates the checksum of the file when reading and writing in order to
+ * detect corrupted files or files that are not up to date
+ */
+export class PPOMStorage {
+  readonly #storageBackend: StorageBackend;
+
+  readonly #readMetadata: () => PPOMFileMetadata;
+
+  readonly #writeMetadata: (metadata: PPOMFileMetadata) => void;
+
+  /**
+   * Creates a PPOMStorage instance.
+   *
+   * @param options - The options passed to the function.
+   * @param options.storageBackend - The storage backend to use for the local storage.
+   * @param options.readMetadata - A function to read the metadata from the local storage.
+   * @param options.writeMetadata - A function to write the metadata to the local storage.
+   */
+  constructor({
+    storageBackend,
+    readMetadata,
+    writeMetadata,
+  }: {
+    storageBackend: StorageBackend;
+    readMetadata: () => PPOMFileMetadata;
+    writeMetadata: (metadata: PPOMFileMetadata) => void;
+  }) {
+    this.#storageBackend = storageBackend;
+    this.#readMetadata = readMetadata;
+    this.#writeMetadata = writeMetadata;
+  }
+
+  /**
+   * Sync the metadata with the version info from the cdn.
+   * 1. Remove the files that are not readable (e.g. corrupted or deleted).
+   * 2. Remove the files that are not in the cdn anymore.
+   * 3. Remove the files that are not up to date in the cdn.
+   * 4. Remove the files that are not in the local storage from the metadata.
+   * 5. Delete the files that are not in the metadata from the local storage.
+   *
+   * @param versionInfo - Version information of metadata files.
+   */
+  async syncMetadata(versionInfo: FileInfo[]): Promise<PPOMFileMetadata> {
+    const metadata = this.#readMetadata();
+    const syncedMetadata = [];
+
+    for (const fileMetadata of metadata) {
+      // check if the file is readable (e.g. corrupted or deleted)
+      try {
+        await this.readFile(fileMetadata.name, fileMetadata.chainId);
+      } catch (exp: any) {
+        continue;
```

Comment:
Do we intentionally just silently ignore any corrupted or deleted files?

Should we insist we have a clean sync and throw here or at the very least log the corrupt file?

Reply:
In the nest loop here we delete all the files which could not be synced: https://github.com/MetaMask/ppom-validator/pull/1/files/2683bf640ee4987bf5a5a4d931c0df5306b2ba0f#diff-c4b0de610ed40436f953144e1813b5060bcb2c2b5766a0fb7a747639e6662759R127

Reply:
I updated code to log error, eslint setup of project does not allow use if `e` or `err`.